lib/spack/spack/solver/asp.py

```diff
@@ -3686,6 +3568,38 @@ def depends_on(
 
         self.reset()
 
+    @staticmethod
+    def node_for(name: str) -> str:
+        return f'node(ID{name.replace("-", "_")}, "{name}")'
+
+    def rule_body_from(self, when_spec: "spack.spec.Spec") -> Tuple[str, str]:
+        """Computes the rule body from a "when" spec, and returns it, along with the
+        node variable.
+        """
+
+        node_placeholder = "XXX"
+        node_variable = "node(ID, Package)"
+        when_substitutions = {}
+        for s in when_spec.traverse(root=False):
+            when_substitutions[f'"{s.name}"'] = self.node_for(s.name)
+        when_spec.name = node_placeholder
+        body_clauses = self._setup.spec_clauses(when_spec, body=True)
+        for clause in body_clauses:
+            if clause.args[0] == "virtual_on_incoming_edges":
+                # Substitute: attr("virtual_on_incoming_edges", ProviderNode, Virtual)
+                # with: attr("virtual_on_edge", ParentNode, ProviderNode, Virtual)
+                # (avoid adding virtuals everywhere, if a single edge needs it)
+                _, provider, virtual = clause.args
+                clause.args = "virtual_on_edge", node_placeholder, provider, virtual
+        body_str = (
+            f"  {f',{os.linesep}  '.join(str(x) for x in body_clauses)},\n"
+            f"  not external({node_variable}),\n"
+            f"  not runtime(Package)"
```

Comment:
```suggestion
            f"  not external({node_variable})"
```

The `compiler-wrapper` package needs to be a dependency of runtimes too (if they are compiled).

Reply:
Let's save this one for when we add `hip` and friends, so we can work out the consequences there. Right now the runtimes we support follow this rule.